### ISSUE TITLE
builtins.fetchTarball breaks under local?root=

### ISSUE THREAD

**Issue by cleverca22** [NixOS/nix#2405]
```
[clever@amd-nixos:~]$ nix-instantiate -E '(import (builtins.fetchTarball https://github.com/nixos/nixpkgs/archive/master.tar.gz) {}).hello' --store local?root=/home/clever/fakeroot/
unpacking 'https://github.com/nixos/nixpkgs/archive/master.tar.gz'...
nix-instantiate: src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.
Aborted (core dumped)
```
related: #2125, #2174 and #2381

**Comment by a-schaefers** [comment]
Hi, I had the same result whilst trying to bootstrap my system which uses home-manager, which uses fetchtarball also. 

```
[0.1/0.0 MiB DL] downloading 'https://github.com/rycee/home-manager/archive/master.tar.gz'nix: src/libe
xpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(c
tx)' failed.                                                                                          
/run/current-system/sw/bin/nixos-install: line 98:  1680 Aborted                 nix build --out-link "
$outLink" --store "$mountPoint" "${extraBuildFlags[@]}" --extra-substituters "$sub" -f '<nixpkgs/nixos>
' system -I "nixos-config=$NIXOS_CONFIG"
```

**Comment by Spass1966** [comment]
Is this bug upstream? Is there an issue opened for it?

**Comment by cleverca22** [comment]
i think you can work around the problem by copying the output of `fetchTarball` from `/nix/store` to `/mnt/nix/store` using `nix copy --to local?root=/mnt/ /nix/store/theoutput`
but the hard part is finding what the output from `fetchTarball` was

**Comment by edolstra** [comment]
Fixed in 06010eaf199005a393f212023ec5e8bc97978537.
